### PR TITLE
Add support for distributed Cameras

### PR DESCRIPTION
**Pull request by AnthonyHorton**
_Moved_ _from_ _https://github.com/panoptes/POCS/pull/547_

Addressing Issue https://github.com/panoptes/POCS/issues/546 by adding scripts and wrappers to enable distributed control of camera arrays using the Pyro4 library.

- [X] Pyro name server startup
- [X] Pyro camera server startup
- [X] Automatic discovery of remote cameras
- [X] Initialise remote cameras, get UIDs
- [X] Initiate exposures on remote cameras
- [X] Transfer files from remote cameras after exposures
- [x] Autofocus of remote cameras
- [x] `take_observation` using remote cameras
- [x] Access remote camera properties
- [x] Set remote camera properties

### CODE REVIEW

File: conf_files/pyro_camera.yaml
Comment (wtgee) [inline conf_files/pyro_camera.yaml]: I almost want this at the top and in all caps. And it will probably still happen. :)

File: huntsman/camera/__init__.py
Comment (wtgee) [inline huntsman/camera/__init__.py]: This isn't different, right? I realize you were probably just copying straying over, but let's just remove from here as the other is still available to us via POCS.

File: huntsman/camera/__init__.py
Comment (wtgee) [inline huntsman/camera/__init__.py]: s/ise/use/

File: huntsman/camera/__init__.py
Comment (wtgee) [inline huntsman/camera/__init__.py]: maybe the word "detected" in here?

File: huntsman/camera/__init__.py
Comment (wtgee) [inline huntsman/camera/__init__.py]: As far as I can tell lines 220-228 are the only lines different from what would be in POCS. Would it work to move these lines into `create_distributed_cameras` (maybe with rename to `create_distributed_cameras_from_config`) and then call the the functions sequentially? Would have to deal with the primary camera preference in some fashion but you already assume here that the distributed should override the local.

Would just be nice to reduce all this to only what we need separately.

File: huntsman/camera/__init__.py
Comment (wtgee) [inline huntsman/camera/__init__.py]: Since I've been doing a lot of the docs stuff lately, I would add the backticks around the `pocs...Camera` here.

File: huntsman/camera/__init__.py
Comment (wtgee) [inline huntsman/camera/__init__.py:143]: This doesn't match what is returned.

File: huntsman/camera/pyro.py
Comment (wtgee) [inline huntsman/camera/pyro.py]: I'm not seeing this anywhere?

File: huntsman/camera/pyro.py
Comment (wtgee) [inline huntsman/camera/pyro.py]: s/Neet/Need/

File: huntsman/camera/pyro.py
Comment (wtgee) [inline huntsman/camera/pyro.py]: You don't want to store `uri` so can can call `connect()` below without uri again? (Or a thin-wrapper `reconnect` in addition?)

File: huntsman/camera/pyro.py
Comment (wtgee) [inline huntsman/camera/pyro.py]: Would we ever want async?

File: huntsman/camera/pyro.py
Comment (wtgee) [inline huntsman/camera/pyro.py]: Would `os.path.normpath` do this?

File: huntsman/camera/pyro.py
Comment (wtgee) [inline huntsman/camera/pyro.py:186]: I still think I would prefer to go the extra setup mile with ssh keys and then just map the directories so that the file transfer is transparent. The rsync works for now and it looks like this line make it easy to switch out (or remove) in the future.

File: huntsman/camera/pyro.py
Comment (wtgee) [inline huntsman/camera/pyro.py]: Can these two lines be compressed?

File: huntsman/camera/pyro.py
Comment (wtgee) [inline huntsman/camera/pyro.py]: Same.

File: huntsman/camera/pyro.py
Comment (wtgee) [inline huntsman/camera/pyro.py:306]: Is everything above just for serialization?

File: huntsman/camera/pyro.py
Comment (wtgee) [inline huntsman/camera/pyro.py:347]: Wait, what? You can delete empty directories like that via `find`? I always thought you would have to send to `xargs` or something.

File: huntsman/camera/pyro.py
Comment (wtgee) [inline huntsman/camera/pyro.py]: Probably should make the file a parameter just in case.

File: huntsman/camera/pyro.py
Comment (wtgee) [inline huntsman/camera/pyro.py]: `huntsman` as user here.

File: huntsman/camera/pyro.py
Comment (wtgee) [inline huntsman/camera/pyro.py:453]: Looks like you need to insure this is a `Quantity` or add.

File: huntsman/camera/pyro.py
Comment (wtgee) [inline huntsman/camera/pyro.py:461]: Probably force `bool` here too.

File: huntsman/tests/conftest.py
Comment (wtgee) [inline huntsman/tests/conftest.py:62]: Maybe a docstring explaining this deeply nest kill structure. :)

File: huntsman/tests/test_camera.py
Comment (wtgee) [inline huntsman/tests/test_camera.py:170]: Still needed here?

File: huntsman/tests/test_observatory.py
Comment (wtgee) [inline huntsman/tests/test_observatory.py]: Looks like this was supposed to change?

File: huntsman/utils/pyro/__init__.py
Comment (wtgee) [inline huntsman/utils/pyro/__init__.py:100]: Can we just run on the localhost if none is given? Or is auto-detection the assumed usage?

File: huntsman/utils/pyro/__init__.py
Comment (wtgee) [inline huntsman/utils/pyro/__init__.py]: `else` seems unnecessary.

File: huntsman/utils/pyro/__init__.py
Comment (wtgee) [inline huntsman/utils/pyro/__init__.py]: s/$POCS/$HUNTSMAN_POCS/

File: huntsman/utils/pyro/__init__.py
Comment (wtgee) [inline huntsman/utils/pyro/__init__.py:151]: Do you need to catch the `KeyboardInterrupt`?

File: scripts/pyro_name_server.py
Comment (wtgee) [inline scripts/pyro_name_server.py:16]: Did we want to enable this by default or do you think not needed?

**Review by wtgee — COMMENTED** [review]
Whew! That was a long review which means that was even longer for you! 

Thanks for getting this moved over. Submitting just as comment for now but can approve if you would rather merge before vacation. 

Seems like there is a bit we could do to eliminate code duplication between here and POCS, but I'm happy to do that after the fact.

File: conf_files/pyro_camera.yaml
Comment (AnthonyHorton) [reply conf_files/pyro_camera.yaml]: A compromise, I've made it all caps but didn't move it :-)

File: huntsman/camera/__init__.py
Comment (AnthonyHorton) [reply huntsman/camera/__init__.py]: Yes, this entire file is just copied over verbatim from the branch at `panoptes/POCS`. A fair bit of unnecessary code duplication that I hadn't made any attempt to reduce yet. Replacing this with an import now.

File: huntsman/camera/__init__.py
Comment (AnthonyHorton) [reply huntsman/camera/__init__.py]: Well, these few lines plus a small change about how the default primary camera gets chosen (I can't assume that there is a camera named `Cam00` so I sort the camera names and choose the first one). But otherwise it is identical to the POCS version.

Yes, it was the awkwardness of keeping track of whether a primary camera has been selected yet that stopped me from making these two entirely separate functions in the first place. I'll have a think about how to work around that.

It doesn't really override a local primary camera with a distributed one, as there will be at most one primary camera specified in the config file so the distributed camera creation will only set the primary camera is the local camera creation hasn't.

File: huntsman/camera/__init__.py
Comment (AnthonyHorton) [reply huntsman/camera/__init__.py]: Good point.

File: huntsman/camera/__init__.py
Comment (AnthonyHorton) [reply huntsman/camera/__init__.py]: Yep.

File: huntsman/camera/__init__.py
Comment (AnthonyHorton) [reply huntsman/camera/__init__.py]: Yes.

File: huntsman/camera/__init__.py
Comment (AnthonyHorton) [reply huntsman/camera/__init__.py:143]: Oops, forgot to add the primary camera return value.

File: huntsman/camera/pyro.py
Comment (AnthonyHorton) [reply huntsman/camera/pyro.py]: Good catch. Must have been using it at some point, but not any more.

File: huntsman/camera/pyro.py
Comment (AnthonyHorton) [reply huntsman/camera/pyro.py]: There's probably a lot of instances of 'Neet' scattered through the Camera classes. Most of them have this same docstring...

File: huntsman/camera/pyro.py
Comment (AnthonyHorton) [reply huntsman/camera/pyro.py]: Probably a good idea, if for no other reason than to order to allow us to later on to defer the `connect()` instead of calling it from within `__init__()`.   That's something we might want to do, because it's the `connect()` that triggers the remote object creation and, for real cameras, the lengthy hardware initialisation process.

File: huntsman/camera/pyro.py
Comment (AnthonyHorton) [reply huntsman/camera/pyro.py]: We do, for the `take_exposure` and `autofocus` methods. I could just use `Thread`s for those, but instead used Pyro's built in asynchronous results functionality.

File: huntsman/camera/pyro.py
Comment (AnthonyHorton) [reply huntsman/camera/pyro.py]: If I'm reading the docs right `os.path.normpath` would replace the `rstrip` (as well as doing some extra tidying if needed) but I still need the `+ '/'`

File: huntsman/camera/pyro.py
Comment (AnthonyHorton) [reply huntsman/camera/pyro.py:186]: Yep, would be easy to change. ssh keys are already set up, otherwise rsync would be asking for passwords here.

File: huntsman/camera/pyro.py
Comment (AnthonyHorton) [reply huntsman/camera/pyro.py]: Looking at is again I see that I set synchronous mode in `connect()`, and set asynchronous mode at the start of both `take_exposure` and `autofocus`, but I don't switch it back to synchronous at the end of those methods. That ought to break the property access methods if called after an exposure. Hmmm.

Comment (AnthonyHorton) [reply huntsman/camera/pyro.py]: Oh, there's a change in naming here too, because `async` is a reserved keyword in Python 3.7. I've changed to the new names.

File: huntsman/camera/pyro.py
Comment (AnthonyHorton) [reply huntsman/camera/pyro.py]: Yes. Yes they can.

File: huntsman/camera/pyro.py
Comment (AnthonyHorton) [reply huntsman/camera/pyro.py]: Ditto.

File: huntsman/camera/pyro.py
Comment (AnthonyHorton) [reply huntsman/camera/pyro.py:306]: Yes, that's me doing my best to ensure either successful serialisation with Pyro's default serialiser, or an informative error. I added most of this today in response to the `np.int64` fiasco!

File: huntsman/camera/pyro.py
Comment (AnthonyHorton) [reply huntsman/camera/pyro.py:347]: Neat trick, huh? I found this is an answer on StackOverflow, of course.

File: huntsman/camera/pyro.py
Comment (AnthonyHorton) [reply huntsman/camera/pyro.py]: With Pyro the recommended approach is to pass the class to the daemon rather than an object, and let the daemon handle object creation. That precludes passing any arguments to it, so I didn't even bother putting any in the constructor here. It is possible to pass an object instead though, which would enable overriding the config file name.

File: huntsman/camera/pyro.py
Comment (AnthonyHorton) [reply huntsman/camera/pyro.py]: Yep.

File: huntsman/camera/pyro.py
Comment (AnthonyHorton) [reply huntsman/camera/pyro.py:453]: The setter in the underlying `_camera` does that.

File: huntsman/camera/pyro.py
Comment (AnthonyHorton) [reply huntsman/camera/pyro.py:461]: Same as above.

File: huntsman/tests/conftest.py
Comment (AnthonyHorton) [reply huntsman/tests/conftest.py:62]: Did have a few problems with these processes being left hanging after running tests....

File: huntsman/tests/test_camera.py
Comment (AnthonyHorton) [reply huntsman/tests/test_camera.py:170]: Hmmm, These Pyro cameras don't emit the warning that the other Camera classes do (shame!), so I don't need the warnings filter. The test is still useful though, just to make sure that trying to start another exposure while one is in progress doesn't break anything.

I just leave it up to Pyro to handle any 'exposure collisions'. I've configured the camera servers to create only a single camera object, to handle remote method calls sequentially ('multiplex' mode), and made the server side `take_exposure` method blocking. If you try to start another exposure while one is in progress Pyro will just put it in a queue until the remote camera is ready.

File: huntsman/tests/test_observatory.py
Comment (AnthonyHorton) [reply huntsman/tests/test_observatory.py]: Yes. This entire file is just copied verbatim from the POCS version right now. There will be a few things that needs changing, and other buts that are redundant with `test_huntsman.py`. Haven't had chance to go through it yet.

File: huntsman/utils/pyro/__init__.py
Comment (AnthonyHorton) [reply huntsman/utils/pyro/__init__.py:100]: Running on localhost is 'special' for Pyro servers (including the name server), it disable the broadcast responder so they can't be reached from other machines on the network. Good for testing with simulators, but not a very useful default otherwise.

I see the auto detection as a convenience, really. It's probably not 100% reliable for complicated networking configurations where the machine has multiple IP address. For operational systems with fixed IPs it would be better to set the name server host manually via config file.

File: huntsman/utils/pyro/__init__.py
Comment (AnthonyHorton) [reply huntsman/utils/pyro/__init__.py]: It does. The only effect here is to convert a value of `port` that evaluates to `False` (e.g. `0`) into `None`.  I'm not sure why I would want that to happen, though.

File: huntsman/utils/pyro/__init__.py
Comment (AnthonyHorton) [reply huntsman/utils/pyro/__init__.py]: Yes.

File: huntsman/utils/pyro/__init__.py
Comment (AnthonyHorton) [reply huntsman/utils/pyro/__init__.py:151]: I tried doing that, but it didn't work as I expected. The `except` block didn't get executed. So I just put all the cleanup in `finally` and that did work.

File: scripts/pyro_name_server.py
Comment (AnthonyHorton) [reply scripts/pyro_name_server.py:16]: I think it should be enabled by default.  With the `finally` in the camera server they will now de-register themselves from the name server if you shut them down, but that doesn't handle more messy situations (network problems, Pi power failure, out of memory errors, etc.).  The name server autoclean is a handy catch-all backup.

**Comment by AnthonyHorton** [comment]
Thanks for the detailed and quick review! I don't want to push through the merge just yet, though. At the very least want to get tests passing, and ideally remove the overlap and inconsistencies between `pocs.tests.test_observatory.py`, `huntsman.tests.test_observatory.py` & `huntsman.tests.test_huntsman.py`.

**Comment by AnthonyHorton** [comment]
Looking at it a bit more I don't think I actually need anything from `test_observatory.py`, apart from `test_pyro_camera`.  So I've moved that into `test_huntsman.py` and removed `test_observatory.py`. Bingo! Passing tests and a diff that's ~400 lines smaller!

File: huntsman/camera/__init__.py
Comment (AnthonyHorton) [reply huntsman/camera/__init__.py]: OK, reduced code duplication a fair bit by importing `pocs.camera.create_cameras_from_config` to handle the local camera creation.

File: huntsman/camera/__init__.py
Comment (AnthonyHorton) [reply huntsman/camera/__init__.py:143]: Now moot because I took the primary camera part out again.

**Comment by AnthonyHorton** [comment]
OK, reduced duplication of code from POCS in `huntsman/camera/__init__.py` and knocked another ~100 lines off the diff. Can I get a re-review, please, @wtgee? I think this might be OK to merge now.

Haven't done anything for those three unchecked items in the the PR description yet, but I think it's OK for them to be separate Issues & PRs.

File: huntsman/camera/__init__.py
Comment (wtgee) [inline huntsman/camera/__init__.py:122]: Minor point, but since it's an OrderedDict do you want to skip the `sorted` and use the first one placed in the dict? I think it comes out the same anyway.

File: huntsman/camera/pyro.py
Comment (wtgee) [inline huntsman/camera/pyro.py:149]: I'm still on the fence about *ever* allowing indefinite waits but it's fine for now. Maybe we need an issue for it.

File: huntsman/camera/pyro.py
Comment (wtgee) [reply huntsman/camera/pyro.py:306]: I assume the trailing `/` is for `rsync`?

File: huntsman/camera/pyro.py
Comment (wtgee) [inline huntsman/camera/pyro.py:342]: There's some weirdness with the path going on. Maybe an example of what the `source` input is here.

File: huntsman/tests/test_camera.py
Comment (wtgee) [inline huntsman/tests/test_camera.py:4]: Do you want to just name this `test_pyro_camera.py` and then we can leave it for only Pyro items (if we ever add other camera types just to Huntsman)?

File: huntsman/tests/test_camera.py
Comment (wtgee) [reply huntsman/tests/test_camera.py:170]: Got it. i was thinking this was testing the collision of multiple cameras (a la the SBIG simultaneous readout issue)

File: huntsman/tests/test_huntsman.py
Comment (wtgee) [inline huntsman/tests/test_huntsman.py:234]: Note https://github.com/panoptes/POCS/pull/679 and the comments.

File: huntsman/utils/pyro/__init__.py
Comment (wtgee) [reply huntsman/utils/pyro/__init__.py]: I don't see this change.

File: huntsman/utils/pyro/__init__.py
Comment (wtgee) [reply huntsman/utils/pyro/__init__.py:151]: Maybe:

```python
try:
    daemon...
except KeyboardInterrupt:
    logger.info('Stopping  name server')
finally:
    # Cleanup
```

Not necessary though.

Comment (wtgee) [reply huntsman/utils/pyro/__init__.py:151]: That would only be so that we could add an *additional* `except Exception` so we could capture when it fails for some other reason. But again, not necessary for this PR.

File: scripts/pyro_camera_server.py
Comment (wtgee) [inline scripts/pyro_camera_server.py:6]: Could we have this script check for a running name server and refuse to start if so? (Can be separate PR)

**Review by wtgee — APPROVED** [review]
Great! Looks good. Thanks for a ton of work to get this done in the last few days. Have a good vacation!

File: huntsman/camera/__init__.py
Comment (AnthonyHorton) [reply huntsman/camera/__init__.py:122]: I guess it depends on whether we want to default to the first camera added (which will be `Cam00` if there are any local cameras) or to the first camera alphabetically...

File: huntsman/camera/pyro.py
Comment (AnthonyHorton) [reply huntsman/camera/pyro.py:149]: At the least we should be enabling these timeouts in the testing and in the default config files to encourage their use. Will revisit that as part of the general issue of timeouts for all the cameras' threaded actions.

File: huntsman/camera/pyro.py
Comment (AnthonyHorton) [reply huntsman/camera/pyro.py:342]: Ah, yes, the directory cleanup follows the file transfer in the call chain, which just passes on the same `source` argument that the file transfer uses, which is in `user@host:/directory/subdirectory/filename` format ready for rysnc.

File: huntsman/tests/test_camera.py
Comment (AnthonyHorton) [reply huntsman/tests/test_camera.py:4]: Inclined to leave as is to emphasise that this is a lightly edited copy of `pocs/tetsts/test_camera.py`, and is supposed to be kept in sync with that. In fact I'll add a comment about that.

File: huntsman/camera/pyro.py
Comment (AnthonyHorton) [reply huntsman/camera/pyro.py:306]: Yes, that's right.

File: huntsman/utils/pyro/__init__.py
Comment (AnthonyHorton) [reply huntsman/utils/pyro/__init__.py]: Fixed another instance of the same thing and missed this one!

File: huntsman/tests/test_huntsman.py
Comment (AnthonyHorton) [reply huntsman/tests/test_huntsman.py:234]: I'll have to return to that one. Not sure I understand the underlying Iissue fully.

File: scripts/pyro_camera_server.py
Comment (AnthonyHorton) [reply scripts/pyro_camera_server.py:6]: The `run_camera_server` function that the script calls does do exactly that. Might be neater to have the script do it instead, though.

**Pull request merged**
